File: packages/lhc/src/domains/thread-view/internal/seam.ts
// Test injection facility (Story 0, FC-0.6): one mechanism, two named
// points, pattern from Epic 02's intake walk hook. Production code carries
// the points as no-ops unless a test installs a hook:
//   - "post-commit-advance": fired by the boundary advance after intake's
//     commit flush, before the advance computes (TC-4.6's failure injection —
//     an installed hook that throws stands in for an advance failure);
//   - "compact-write": fired by compact between the sweep and the view-write
//     transaction (TC-2.4's crash injection — a throwing hook proves the
//     prior view survives an aborted write).
// Tests reach the setters through test/fixtures/view-seam.ts (the one
// directory sanctioned to import below the SDK surface); production code
// only ever fires.
export type ViewInjectionPoint = "post-commit-advance" | "compact-write";

export type ViewInjectionHook = () => void;

const hooks: Record<ViewInjectionPoint, ViewInjectionHook | null> = {
  "post-commit-advance": null,
  "compact-write": null,
};

export function setViewInjectionHook(
  point: ViewInjectionPoint,
  hook: ViewInjectionHook | null,
): void {
  hooks[point] = hook;
}

// The production-side call: a no-op when nothing is installed. An installed
// hook's throw propagates to the call site on purpose — that is the injected
// failure the call site must survive.
export function fireViewInjection(point: ViewInjectionPoint): void {
  hooks[point]?.();
}

File: packages/lhc-py/src/lhc/thread_view/internal/seam.py
"""Ported from packages/lhc/src/thread-view/internal/seam.ts.

Test injection facility for compact's write path. Production code carries
the point as a no-op unless a test installs a hook: "compact-write" fires
between the sweep and the view-write transaction.
Tests reach the setters through test/fixtures/view-seam.ts (the one
directory sanctioned to import below the SDK surface); production code
only ever fires.
"""

from __future__ import annotations

from collections.abc import Callable
from typing import Literal

ViewInjectionPoint = Literal["compact-write"]

ViewInjectionHook = Callable[[], None]

_hooks: dict[ViewInjectionPoint, ViewInjectionHook | None] = {
    "compact-write": None,
}


def set_view_injection_hook(point: ViewInjectionPoint, hook: ViewInjectionHook | None) -> None:
    _hooks[point] = hook


# The production-side call: a no-op when nothing is installed. An installed
# hook's throw propagates to the call site on purpose — that is the injected
# failure the call site must survive.
def fire_view_injection(point: ViewInjectionPoint) -> None:
    hook = _hooks[point]
    if hook is not None:
        hook()
